src/esp/gfx/BatchRenderer.h

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#ifndef ESP_GFX_BATCH_RENDERER_H_
+#define ESP_GFX_BATCH_RENDERER_H_
+
+#include <Corrade/Containers/Pointer.h>
+#include <Magnum/GL/GL.h>
+#include <Magnum/Magnum.h>
+#include <cstddef>
+
+namespace esp {
+namespace gfx {
+
+enum class BatchRendererFlag {
+  NoTextures = 1 << 0
+  // TODO memory-map
+};
+typedef Corrade::Containers::EnumSet<BatchRendererFlag> BatchRendererFlags;
+CORRADE_ENUMSET_OPERATORS(BatchRendererFlags)
+
+struct BatchRendererConfiguration {
+  explicit BatchRendererConfiguration();
+  ~BatchRendererConfiguration();
+
+  BatchRendererConfiguration& setFlags(BatchRendererFlags flags);
+  BatchRendererConfiguration& setTileSizeCount(
+      const Magnum::Vector2i& tileSize,
+      const Magnum::Vector2i& tileCount);
+
+  struct State;
+  Corrade::Containers::Pointer<State> state;
+};
+
+class BatchRenderer {
+ public:
+  explicit BatchRenderer(const BatchRendererConfiguration& configuration)
+      : BatchRenderer{Magnum::NoCreate} {
+    create(configuration);
+  }
+
+  ~BatchRenderer();
+
+  Magnum::Vector2i tileSize() const;
+  Magnum::Vector2i tileCount() const;
+  /* Same as tileCount().product() */
+  std::size_t sceneCount() const;
+
+  // TODO take an importer instead? that way the consumer can configure it
+  void addFile(Corrade::Containers::StringView filename);
+  void addFile(Corrade::Containers::StringView filename,
+               Corrade::Containers::StringView importerPlugin);
+
+  // TODO "if there's a scene, name corresponds to a root bject name,
+  //  otherwise it's the whole file"
+  // TODO or use empty name for the whole scene? but which file??? or use a
+  //  scene name for the whole scene??
+  // TODO document this may not return consecutive IDs in case the added name
+  //  is multiple meshes together
+  std::size_t add(Magnum::UnsignedInt sceneId,
+                  Corrade::Containers::StringView name);
+  std::size_t add(Magnum::UnsignedInt sceneId,
```

Comment:
We've also used the term "node template", i.e. the name you give here is the name of a node in the composite gltf, and it may have a hierarchy with multiple meshes. What's the status here? Does this return an integer representing the entire instantiated node? And we use that integer to index into `transformations` below?

Reply:
Yes, that's what this does. (The TODO comments may be confusing, there I was just noting down ideas for how to solve naming when multiple files get involved.)

As a proof, look at [the `meshHierarchy()` test case](https://github.com/facebookresearch/habitat-sim/blob/63f4e24a605bb5d4f9afcae3cb8ea8f9df7cac26/src/tests/BatchRendererTest.cpp#L416-L462) -- it corresponds to [this portion of the glTF file](https://github.com/facebookresearch/habitat-sim/blob/63f4e24a605bb5d4f9afcae3cb8ea8f9df7cac26/data/test_assets/scenes/batch.gltf#L206-L265), and this is the rendered output:

![BatchRendererTestMeshHierarchy.png](https://raw.githubusercontent.com/facebookresearch/habitat-sim/63f4e24a605bb5d4f9afcae3cb8ea8f9df7cac26/data/test_assets/screenshots/BatchRendererTestMeshHierarchy.png)

What this doesn't currently handle is deeper hierarchies, as I felt that was non-essential if you have full control over the input data and don't need to modify transformations inside the hierarchy. But I expect this to become a use case when generic input files are allowed, and it's prepared for that.